machine_learning/neural_network.cpp

```diff
@@ -0,0 +1,719 @@
+                        std::cout << "Training: Epoch " << epoch << '/' << epochs; 
+                        std::cout << ", Loss: " << loss;
+                        std::cout << ", Accuracy: " << acc;
+                        std::cout << ", Taken time: " << duration.count() / 1e6 << " seconds";
+                        std::cout << std::endl;
+                    }
+                    return;
+                }
+
+                /**
+                 * Function to fit model on data stored in csv file
+                 * @param file_name csv file name
+                 * @param last_label flag for whether label is in first or last column
+                 * @param epochs number of epochs
+                 * @param learning_rate learning rate 
+                 * @param normalize flag for whether to normalize data 
+                 * @param batch_size batch size for gradient descent (default = 32)
+                 * @param shuffle flag for whether to shuffle data
+                 */                
+                void fit_from_csv (const std::string &file_name, 
+                                     const bool &last_label,
+                                     const int &epochs, 
+                                     const double &learning_rate,
+                                     const bool &normalize,
+                                     const size_t &batch_size = 32,
+                                     const bool &shuffle = true) {
+                    // Getting training data from csv file                    
+                    auto data = this -> get_XY_from_csv(file_name, last_label, normalize);
+                    // Fit the model on training data
+                    this -> fit(data.first, data.second, epochs, learning_rate, batch_size, shuffle);
+                    return;
+                }
+
+                /**
+                 * Function to evaluate model on supplied data
+                 * @param X X data (input data)
+                 * @param Y Y data (label)
+                 */  
+                void evaluate(const std::vector< std::vector <std::valarray <double>>> &X, 
+                                const std::vector< std::vector <std::valarray <double>>> &Y) {
+                    std::cout << "INFO: Evaluation Started" << std::endl;
+                    double acc = 0, loss = 0; // intialize performance metrics with zero
+                    for(size_t i = 0; i < X.size(); i++) { // For every sample in input
+                        // Get predictions
+                        std::vector<std::valarray<double>> pred = this -> single_predict(X[i]);
+                        // If predicted class is correct 
+                        if(argmax(pred) == argmax(Y[i])) {
+                            acc += 1; // Increment accuracy
+                        }
+                        // Calculating loss - Mean Squared Error
+                        loss += sum(apply_function((Y[i] - pred), 
```

Comment:
`Y[I]` is a single value and `apply_function` seems redundant. Maybe it can be moved outside the for-loop as a one-time computation?

Reply:
Actually both ```Y[i]``` and ```pred``` are a ```2D```vectors. They will have shape ```(1, 1)``` in case of regression and ```(1, X)``` in case of classification (X = number of classes).

Reply:
ah, ok